horovod/common/ops/gloo_operations.cc

```diff
@@ -0,0 +1,264 @@
+// Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "gloo_operations.h"
+
+namespace horovod {
+  namespace common {
+
+    IGlooAlgorithms *GetAlgorithmsForType(DataType dtype, GlooContext *gloo_context) {
+      switch (dtype) {
+        case HOROVOD_UINT8:
+          return new GlooAlgorithms<u_int8_t>(gloo_context);
+        case HOROVOD_INT8:
+          return new GlooAlgorithms<int8_t>(gloo_context);
+        case HOROVOD_UINT16:
+          return new GlooAlgorithms<u_int16_t>(gloo_context);
+        case HOROVOD_INT16:
+          return new GlooAlgorithms<int16_t>(gloo_context);
+        case HOROVOD_INT32:
+          return new GlooAlgorithms<int32_t>(gloo_context);
+        case HOROVOD_INT64:
+          return new GlooAlgorithms<int64_t>(gloo_context);
+        case HOROVOD_FLOAT16:
+          return new GlooAlgorithms<gloo::float16>(gloo_context);
+        case HOROVOD_FLOAT32:
+          return new GlooAlgorithms<float>(gloo_context);
+        case HOROVOD_FLOAT64:
+          return new GlooAlgorithms<double>(gloo_context);
+        case HOROVOD_BOOL:
+          return new GlooAlgorithms<bool>(gloo_context);
+        default:
+          throw std::logic_error("Type " + DataType_Name(dtype) +
+                                 " is not supported in Gloo mode.");
+      }
+    }
+
+    template<typename T>
+    GlooAlgorithms<T>::GlooAlgorithms(GlooContext *gloo_context)
+        : gloo_context_(gloo_context) {}
+
+
+    template<typename T>
+    void GlooAlgorithms<T>::Allreduce(void *buffer_data, int num_elements) {
+      gloo::AllreduceOptions opts(gloo_context_->ctx);
+      opts.setOutput((T *) buffer_data, num_elements);
+
+      // set allreduce function
+      opts.setReduceFunction([](void *a, const void *b, const void *c, size_t n) {
```

Comment:
I'm wondering if there's a cleaner way to do this.  Maybe:

```
void (*func)(void*, const void*, const void*, size_t) = &::gloo::sum<T>; opts.setReduceFunction(AllreduceOptions::Func(func));
```

As shown here: https://github.com/facebookincubator/gloo/blob/master/gloo/test/allreduce_test.cc#L370

What do you think?

Reply:
Yes. Thanks for the remind !